src/encoder/ffmpeg_encoder.ml

```diff
@@ -0,0 +1,127 @@
+(*****************************************************************************
+
+  Liquidsoap, a programmable audio stream generator.
+  Copyright 2003-2019 Savonet team
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details, fully stated in the COPYING
+  file at the root of the liquidsoap distribution.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
+
+ *****************************************************************************)
+
+(** FFMPEG encoder *)
+
+open FFmpeg
+
+module Resampler = Swresample.Make (Swresample.PlanarFloatArray) (Swresample.Frame)
+
+let log = Ffmpeg_config.log
+
+type handler = {
+  output: Avutil.output Avutil.container;
+  stream: (Avutil.output, Avutil.audio) Av.stream;
+  converter: (Swresample.PlanarFloatArray.t, Swresample.Frame.t) Swresample.ctx
+}
+
+let encoder ffmpeg meta =
+  let short_name = ffmpeg.Ffmpeg_format.format in
+  let format =
+    match Av.Format.guess_output_format ~short_name () with
+      | None -> failwith "No format for filename!"
+      | Some f -> f
+  in
+  let codec_id =
+    Avcodec.Audio.find_id ffmpeg.Ffmpeg_format.codec
+  in
+  let out_sample_format =
+    Avcodec.Audio.find_best_sample_format codec_id `Dbl
+  in
+  let src_freq = Frame.audio_of_seconds 1. in
+  let channels = Lazy.force Frame.audio_channels in
+  let src_channels =
+    match channels with
+      | 1 -> `Mono
+      | 2 -> `Stereo
+      | _ -> failwith "%ffmpeg encoder only supports mono or stereo audio for now!"
+  in
+  let dst_freq = 
+    Lazy.force ffmpeg.Ffmpeg_format.samplerate
+  in
+  let dst_channels =
+    match ffmpeg.Ffmpeg_format.channels with
+      | 1 -> `Mono
+      | 2 -> `Stereo
+      | _ -> failwith "%ffmpeg encoder only supports mono or stereo audio for now!"
+  in
+  let buf = Buffer.create 1024 in
+  let make () =
+    let converter =
+      Resampler.create ~out_sample_format
```

Comment:
Can't we use ffmpeg to do the resampling here?

Reply:
I think it's doing it no?

Reply:
isn't `Resampler` using libsamplerate by default?

Reply:
In this context, `Resampler` is a module generated via `Swresample` above that converts to the native `frame` format for `FFmpeg`.

Reply:
Ah ok then, sorry!